[in_app_purchase] Add support for promotional offers through Store-Kit wrappers

mvanbeusekom opened the pull request:
This PR adds support for [promotional offers](https://developer.apple.com/documentation/storekit/original_api_for_in-app_purchase/subscriptions_and_offers/implementing_promotional_offers_in_your_app?language=objc) on iOS by exposing the `SKProduct.discounts` (via the `SKProductWrapper` Dart class) and the `SKMutablePayment.paymentDiscount` (via the `SKPaymentWrapper` Dart class) properties. 

Resolves https://github.com/flutter/flutter/issues/92231

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Note that unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I [updated pubspec.yaml](https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates) with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code

cyanglaz commented on packages/in_app_purchase/in_app_purchase_ios/ios/Classes/FIAObjectTranslator.h:
```diff
@@ -47,6 +51,9 @@ NS_ASSUME_NONNULL_BEGIN
                  andSKPaymentTransaction:(SKPaymentTransaction *)transaction
     API_AVAILABLE(ios(13), macos(10.15), watchos(6.2));
 
+// Creates an instance of the SKPaymentDiscount class based on the supplied disctionary.
```
nits: dictionary
```suggestion
// Creates an instance of the SKPaymentDiscount class based on the supplied dictionary.
```

mvanbeusekom replied:
Fixed.

cyanglaz commented on packages/in_app_purchase/in_app_purchase_ios/ios/Classes/FIAObjectTranslator.m:
```diff
@@ -24,6 +24,7 @@ + (NSDictionary *)getMapFromSKProduct:(SKProduct *)product {
   // https://github.com/flutter/flutter/issues/26610
   [map setObject:[FIAObjectTranslator getMapFromNSLocale:product.priceLocale] ?: [NSNull null]
           forKey:@"priceLocale"];
+
```
nits: revert

mvanbeusekom replied:
Fixed.

cyanglaz commented on packages/in_app_purchase/in_app_purchase_ios/ios/Classes/FIAObjectTranslator.m:
```diff
@@ -49,6 +54,17 @@ + (NSDictionary *)getMapFromSKProductSubscriptionPeriod:(SKProductSubscriptionPe
 
++ (nonnull NSArray *)getMapArrayFromSKProductDiscounts:
+    (nonnull NSArray<SKProductDiscount *> *)productDiscounts {
+  NSMutableArray *discountsMapArray = [NSMutableArray new];
```
use `alloc] init];` instead of `new`

mvanbeusekom replied:
Fixed.

cyanglaz commented on packages/in_app_purchase/in_app_purchase_ios/ios/Classes/FIAObjectTranslator.m:
```diff
@@ -193,4 +209,24 @@ + (NSDictionary *)getMapFromSKStorefront:(SKStorefront *)storefront
   return map;
 }
 
++ (SKPaymentDiscount *)getSKPaymentDiscountFromMap:(NSDictionary *)map {
```
Is this method used?

mvanbeusekom replied:
No it was not used, however it should have been used. I have added usage (including tests) in commit 6abe36d.

cyanglaz commented on packages/in_app_purchase/in_app_purchase_ios/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart:
```diff
@@ -450,6 +452,13 @@ class SKPaymentWrapper {
   /// testing.
   final bool simulatesAskToBuyInSandbox;
 
+  /// The details of optional discount that should be applied to the payment.
```
Can we have a little more details here?
[Setting Up Promotional Offers] should cover most of the business logic on how to use this variable. We should also add more details on technical part.

For example, when will this value to be `null`

mvanbeusekom replied:
Fixed.

cyanglaz commented on packages/in_app_purchase/in_app_purchase_ios/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart:
```diff
@@ -473,3 +482,99 @@ class SKPaymentWrapper {
+/// Used to indicate a discount is applicable to a payment. The
+/// [SKPaymentDiscountWrapper] instance should be assigned to the
+/// [SKPaymentWrapper] object to which the discount should be applied.
+/// Discount offers are set up in App Store Connect.
```
Maybe offer a link for how to set up the discount offer? (if any)

mvanbeusekom replied:
Done.

cyanglaz commented on packages/in_app_purchase/in_app_purchase_ios/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart:
```diff
@@ -473,3 +482,99 @@ class SKPaymentWrapper {
+  /// Constructs an instance of this from a key value map of data.
+  ///
+  /// The map needs to have named string keys with values matching the names and
+  /// types of all of the members on this class. The `map` parameter must not be
```
nits: "The `map` parameter must not be null." is not necessary as it is a NNBD supported file.

mvanbeusekom replied:
Removed.

cyanglaz commented on packages/in_app_purchase/in_app_purchase_ios/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart:
```diff
@@ -473,3 +482,99 @@ class SKPaymentWrapper {
+
+  /// A string identifying the key that is used to generate the signature.
+  ///
+  /// Keys are generated and downloaded from App Store Connect. See the "KEY ID"
```
Maybe a link to some official guide might help explain this better? I feel like this is hard to maintain. For example, if App Store Connect changed their UI, this has to be updated to match.

mvanbeusekom replied:
Done.

cyanglaz commented on packages/in_app_purchase/in_app_purchase_ios/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart:
```diff
@@ -473,3 +482,99 @@ class SKPaymentWrapper {
+
+  /// A universal unique identifier created together with the signature.
+  ///
+  /// The unique nonce should be generated on your server when it creates the
+  /// `signature` for the payment discount. The nonce can be used once, a new
+  /// nonce should be created for each payment request.
+  /// The string representation of the nonce must be lowercase.
```
Is there an official guide on how to generate this from Apple, if so, we can add a link here too.

mvanbeusekom replied:
Yes there is, I have added an extra link.

cyanglaz commented on packages/in_app_purchase/in_app_purchase_ios/lib/src/store_kit_wrappers/sk_product_wrapper.dart:
```diff
@@ -241,7 +241,8 @@ class SKProductWrapper {
     this.subscriptionPeriod,
     this.introductoryPrice,
-  });
+    List<SKProductDiscountWrapper>? discounts,
```
```suggestion
    this.discounts = <SKProductDiscountWrapper>[],
```
And remove the next line.

cyanglaz replied:
I remember there's a way to indicate default value to be empty array in jsonSerializable.

mvanbeusekom replied:
Updating `List<SKProductDiscountWrapper>? discounts,` to `this.discounts = <SKProductDiscountWrapper>[],` will not work as the default value is not a constant. The following error is generated:

```
The default value of an optional parameter must be constant. dart(non_constant_default_value)
```

That is why I resolved it like this. The field is already configured to use an empty array as default value in jsonSerializable, however this is done on field level not in the constructor:

```dart
  @JsonKey(defaultValue: <SKProductDiscountWrapper>[])
  final List<SKProductDiscountWrapper> discounts;
```

I am not aware you can add additional attributes on constructor parameters.

cyanglaz replied:
now you've updated the filed to be none-optional, will `this.discounts = const <SKProductDiscountWrapper>[],` not work? The parameter doesn't have to be optional, right?

cyanglaz commented on packages/in_app_purchase/in_app_purchase_ios/lib/src/store_kit_wrappers/sk_product_wrapper.dart:
```diff
@@ -295,6 +296,16 @@ class SKProductWrapper {
+  /// An array of subscription offers available for the auto-renewable subscription (available on iOS 12.2 and higher).
+  ///
+  /// This property lists all promotional offers set up in App Store Connect. If
+  /// no promotional offers have been set up this field returns an empty list.
```
nits: 
```suggestion
  /// no promotional offers have been set up, this field returns an empty list.
```

mvanbeusekom replied:
Fixed.

cyanglaz reviewed (changes requested):
Thanks for the quick PR! I left some comments and questions.

mvanbeusekom commented on packages/in_app_purchase/in_app_purchase_ios/example/ios/RunnerTests/Stubs.m:
```diff
@@ -61,6 +61,14 @@ - (instancetype)initWithMap:(NSDictionary *)map {
               forKey:@"subscriptionGroupIdentifier"];
     }
+    if (@available(iOS 12.2, *)) {
+      NSMutableArray *discounts = [NSMutableArray new];
```
Fixed.

mvanbeusekom commented on packages/in_app_purchase/in_app_purchase_ios/example/ios/RunnerTests/TranslatorTests.m:
```diff
@@ -14,6 +14,7 @@ @interface TranslatorTest : XCTestCase
 @property(strong, nonatomic) NSMutableDictionary *productMap;
 @property(strong, nonatomic) NSDictionary *productResponseMap;
 @property(strong, nonatomic) NSDictionary *paymentMap;
+@property(strong, nonatomic) NSDictionary *paymentDiscountMap;
```
Fixed.

cyanglaz commented on packages/in_app_purchase/in_app_purchase_ios/ios/Classes/FIAObjectTranslator.m:
```diff
@@ -193,4 +208,24 @@ + (NSDictionary *)getMapFromSKStorefront:(SKStorefront *)storefront
+    return nil;
+  }
+
+  NSString *identifier = map[@"identifier"];
```
what happens if the `map` is empty, or some of the values are NSNull?

mvanbeusekom replied:
I haven't tested on that since on the Dart side all fields are mandatory. Since this is only called from Dart side I skipped doing additional tests.

If you feel it is better to do so I will expand the logic.

cyanglaz replied:
Thanks! the asserts are for testings, for prod app, should we also prevent nil values?

mvanbeusekom commented:
@cyanglaz thanks for the review. I have added `NSAssert` statements to validate if all required `paymentDiscount` fields are supplied. I have also added additional unit tests to ensure the asserts are there (so somebody removing them will be warned).

cyanglaz commented on packages/in_app_purchase/in_app_purchase_ios/ios/Classes/FIAObjectTranslator.m:
```diff
@@ -193,4 +208,34 @@ + (NSDictionary *)getMapFromSKStorefront:(SKStorefront *)storefront
+  NSString *signature = map[@"signature"];
+  NSNumber *timestamp = map[@"timestamp"];
+
+  SKPaymentDiscount *discount = [[SKPaymentDiscount alloc] initWithIdentifier:identifier
```
what happens to the `SKPaymentDiscount` object if any of the values are nil? Will it crash or throw some iap server error?

mvanbeusekom replied:
If the `SKPaymentDiscount` object it self is `nil` the payment will continue without discount. This is also documented as part of the Dart API.

The values of a `SKPaymentDiscount` should never be `nil` as this would indicate an invalid discount offer. Apples documentation is not very clear on this but it should result into a server error. 

I have restructured the code and added additional validation to ensure empty and `nil` values are detected and returns a `PlatformError` in case invalid values are supplied. This to make it more clear to developers what went wrong.

cyanglaz commented on packages/in_app_purchase/in_app_purchase_ios/ios/Classes/FIAObjectTranslator.m:
```diff
@@ -193,4 +208,34 @@ + (NSDictionary *)getMapFromSKStorefront:(SKStorefront *)storefront
+
+  NSString *identifier = map[@"identifier"];
+  NSString *keyIdentifier = map[@"keyIdentifier"];
+  NSUUID *nonce = [[NSUUID alloc] initWithUUIDString:map[@"nonce"]];
```
will `initWithUUIDString` crash if map[@"nonce"] is nil?

mvanbeusekom replied:
According to [Apple's documentation](https://developer.apple.com/documentation/foundation/nsuuid/1411662-initwithuuidstring?language=objc) the `initWithUUIDString` method returns `nil` for invalid strings. I have made a quick test and confirmed this also includes `nil` values passed in.

I did restructure the code to improve error handling (see earlier comment).

mvanbeusekom commented:
Hi @cyanglaz, 

Last week I processed your feedback but forgot to request another review. Would you mind having another look? Many thanks in advance.

cyanglaz reviewed (approved):
LGTM
